src/traces/scattermapbox/plot.js

```diff
@@ -56,72 +75,108 @@ proto.setSourceData = function(k, opts) {
 };
 
 proto.addLayer = function(k, opts, below) {
-    this.subplot.addLayer({
-        type: k,
+    var source = {
+        type: opts.type,
         id: this.layerIds[k],
         source: this.sourceIds[k],
         layout: opts.layout,
-        paint: opts.paint
-    }, below);
+        paint: opts.paint,
+    };
+    if(opts.filter) {
+        source.filter = opts.filter;
+    }
+    this.subplot.addLayer(source, below);
 };
 
 proto.update = function update(calcTrace) {
+    var trace = calcTrace[0].trace;
     var subplot = this.subplot;
     var map = subplot.map;
     var optsAll = convert(subplot.gd, calcTrace);
     var below = subplot.belowLookup['trace-' + this.uid];
-    var i, k, opts;
-
-    if(below !== this.below) {
-        for(i = ORDER.length - 1; i >= 0; i--) {
-            k = ORDER[i];
+    var i, k, opts, order;
+
+    if(trace.cluster.enabled === this.clusterEnabled) {
```

Comment:
To make this part more readable, I suggest declaring two variables like this and use them everywhere below.

```js
var hasCluster = trace.cluster.enabled;
var hadCluster = this.clusterEnabled;
if(hasCluster === hadCluster) {
```